x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentRebalancer.java

```diff
@@ -0,0 +1,286 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.inference.assignment;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.elasticsearch.ResourceAlreadyExistsException;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.unit.ByteSizeValue;
+import org.elasticsearch.xpack.core.ml.action.StartTrainedModelDeploymentAction;
+import org.elasticsearch.xpack.core.ml.inference.assignment.RoutingInfo;
+import org.elasticsearch.xpack.core.ml.inference.assignment.RoutingState;
+import org.elasticsearch.xpack.core.ml.inference.assignment.TrainedModelAssignment;
+import org.elasticsearch.xpack.ml.MachineLearning;
+import org.elasticsearch.xpack.ml.inference.assignment.planning.AssignmentPlan;
+import org.elasticsearch.xpack.ml.inference.assignment.planning.AssignmentPlanner;
+import org.elasticsearch.xpack.ml.job.NodeLoad;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.OptionalInt;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.stream.Collectors;
+
+class TrainedModelAssignmentRebalancer {
+
+    private static final Logger logger = LogManager.getLogger(TrainedModelAssignmentRebalancer.class);
+
+    private final TrainedModelAssignmentMetadata currentMetadata;
+    private final Map<DiscoveryNode, NodeLoad> nodeLoads;
+    private final Optional<StartTrainedModelDeploymentAction.TaskParams> modelToAdd;
+
+    TrainedModelAssignmentRebalancer(
+        TrainedModelAssignmentMetadata currentMetadata,
+        Map<DiscoveryNode, NodeLoad> nodeLoads,
+        Optional<StartTrainedModelDeploymentAction.TaskParams> modelToAdd
+    ) {
+        this.currentMetadata = Objects.requireNonNull(currentMetadata);
+        this.nodeLoads = Objects.requireNonNull(nodeLoads);
+        this.modelToAdd = Objects.requireNonNull(modelToAdd);
+    }
+
+    TrainedModelAssignmentMetadata.Builder rebalance() {
+        TrainedModelAssignmentMetadata.Builder builder = TrainedModelAssignmentMetadata.Builder.fromMetadata(currentMetadata);
+        if (modelToAdd.isPresent() && builder.hasModel(modelToAdd.get().getModelId())) {
+            throw new ResourceAlreadyExistsException("assignment for model with id [{}] already exists", modelToAdd.get().getModelId());
+        }
```

Comment:
I was initially worried that we would't catch this error  anywhere, but I think we are ok double checking all the places in ClusterState, I just don't want this to be an unhandled exception within our threads that ends up causing issues :/